fvm/evm/offchain/utils/collection_test.go

```diff
@@ -2,45 +2,252 @@ package utils_test
+// `allocators_<height>.gob` checkpoint files for each height. If these checkpoint files exist,
+// the corresponding event JSON files will be skipped to optimize replay.
+func BackwardCompatibleSinceEVMGenesisBlock(
 	t *testing.T,
 	chainID flow.ChainID,
-	storage types.BackendStorage,
-	filePath string,
+	eventsFilePaths []string, // ordered EVM events in JSONL format
+	checkpointDir string,
+	checkpointEndHeight uint64, // EVM height of an EVM state that a checkpoint was created for
 ) {
+	// ensure that checkpoints are not more than the event files
+	require.True(t, len(eventsFilePaths) > 0)
+
+	log.Info().Msgf("replaying EVM events from %v to %v, with checkpoints in %s, and checkpointEndHeight: %v",
+		eventsFilePaths[0], eventsFilePaths[len(eventsFilePaths)-1],
+		checkpointDir, checkpointEndHeight)
+
+	store, checkpointEndHeightOrZero := initStorageWithCheckpoints(t, chainID, checkpointDir, checkpointEndHeight)
+
+	// the events to replay
+	nextHeight := checkpointEndHeightOrZero + 1
+
+	// replay each event files
+	for _, eventsFilePath := range eventsFilePaths {
+		log.Info().Msgf("replaying events from %v, nextHeight: %v", eventsFilePath, nextHeight)
+
+		checkpointEndHeight := replayEvents(t, chainID, store, eventsFilePath, checkpointDir, nextHeight)
+		nextHeight = checkpointEndHeight + 1
+	}
 
+	log.Info().
+		Msgf("succhessfully replayed all events and state changes are consistent with onchain state change. nextHeight: %v", nextHeight)
+}
+
+func initStorageWithCheckpoints(t *testing.T, chainID flow.ChainID, checkpointDir string, checkpointEndHeight uint64) (
+	*TestValueStore, uint64,
+) {
 	rootAddr := evm.StorageAccountAddress(chainID)
 
-	// setup the rootAddress account
-	as := environment.NewAccountStatus()
-	err := storage.SetValue(rootAddr[:], []byte(flow.AccountStatusKey), as.ToBytes())
+	// if there is no checkpoint, create a empty store and initialize the account status,
+	// return 0 as the genesis height
+	if checkpointEndHeight == 0 {
+		store := GetSimpleValueStore()
+		as := environment.NewAccountStatus()
+		require.NoError(t, store.SetValue(rootAddr[:], []byte(flow.AccountStatusKey), as.ToBytes()))
+
+		return store, 0
+	}
+
+	valueFileName, allocatorFileName := checkpointFileNamesByEndHeight(checkpointDir, checkpointEndHeight)
+	values, err := deserialize(valueFileName)
 	require.NoError(t, err)
+	allocators, err := deserializeAllocator(allocatorFileName)
+	require.NoError(t, err)
+	store := GetSimpleValueStorePopulated(values, allocators)
+	return store, checkpointEndHeight
+}
 
-	bp, err := blocks.NewBasicProvider(chainID, storage, rootAddr)
+func replayEvents(
+	t *testing.T,
+	chainID flow.ChainID,
+	store *TestValueStore, eventsFilePath string, checkpointDir string, initialNextHeight uint64) uint64 {
+
+	rootAddr := evm.StorageAccountAddress(chainID)
+
+	bpStorage := storage.NewEphemeralStorage(store)
+	bp, err := blocks.NewBasicProvider(chainID, bpStorage, rootAddr)
```

Comment:
```suggestion
	bp, err := blocks.NewBasicProvider(chainID, store, rootAddr)
```

Reply:
The above will also remove the need to manually update the block hash list, as we do below with:
```go
// commit all block hash list changes
for k, v := range bpStorage.StorageRegisterUpdates() {
	err = store.SetValue([]byte(k.Owner), []byte(k.Key), v)
	require.NoError(t, err)
}
```
Or am I missing something about this? :thinking:

Reply:
But if we want to verify with Execution Data form an EN or AN snapshot, we need to move this tests to a util. Probably extract the common part to be reused by testcase or util.